components/script/dom/htmlelement.rs

```diff
@@ -421,11 +424,68 @@ impl HTMLElementMethods for HTMLElement {
     }
 
     // https://html.spec.whatwg.org/multipage/#the-innertext-idl-attribute
-    fn SetInnerText(&self, _: DOMString) {
-        // XXX (ferjm) implement this.
+    fn SetInnerText(&self, input: DOMString) {
+        // Step 1.
+        let document = document_from_node(self);
+
+        // Step 2.
+        let fragment = DocumentFragment::new(&document);
+
+        // Step 3. The given value is already named 'input'.
+
+        // Step 4.
+        let mut position = input.chars().peekable();
+
+        // Step 5.
+        let mut text = String::new();
+
+        // Step 6.
+        loop {
+            match position.next() {
+                Some(ch) => match ch {
+                    '\u{000A}' | '\u{000D}' => {
+                        if ch == '\u{000D}' && position.peek() == Some(&'\u{000A}') {
+                            // a \r\n pair should only generate one <br>,
+                            // so just skip the \r.
+                            position.next();
+                        }
+
+                        if !text.is_empty() {
+                            append_text_node_to_fragment(&document, &fragment, &text);
+                            text.clear();
+                        }
+
+                        let br = HTMLBRElement::new(local_name!("br"), None, &document);
+                        let node = DomRoot::upcast::<Node>(br);
+                        fragment.upcast::<Node>().AppendChild(&node).unwrap();
+                    },
+                    _ => {
+                        text.push(ch);
+                    }
+                },
+                None => {
+                    if !text.is_empty() {
+                        append_text_node_to_fragment(&document, &fragment, &text);
+                        text.clear();
+                    }
+                    break
+                }
+            }
+        }
+
+        // Step 7.
+        Node::replace_all(Some(fragment.upcast()), self.upcast::<Node>());
     }
 }
 
+fn append_text_node_to_fragment(document: &Document,
+                                fragment: &DocumentFragment,
+                                text: &str) {
```

Comment:
nit: It may be clearer, and require less allocations, to take `text` as a `String`. You should be able to assign `text = String::new()` after calling this function, or use `mem::swap` / `mem::replace` to replace them by an empty string.

Reply:
BTW, I changed this, but I don't really understand how the previous version required more allocations. Would you mind explaining it, please? The previous version is https://gist.github.com/ferjm/eae14994a02d2867e0b31b452b2d84fc